torch/nn/modules/activation.py

```diff
@@ -1038,7 +1038,16 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
         """
         is_batched = query.dim() == 3
         if self.batch_first and is_batched:
-            query, key, value = [x.transpose(1, 0) for x in (query, key, value)]
+            if key is value:
```

Comment:
This will make a difference in the _in_projection_packed method in torch/nn/functional.py. When we make transpose for q,k,v respectively, it will never satisfy the conditions of self-attention (needing "key is value" and "query is key") or encoder-decoder attention (needing "key is value" and "not query is key"). It's true it should not affect the values. However we observed if we need 3 linear module instantiations, it can take more time in the forward/backward processes (especially when we use mix-precision training, the effect of extra casting also slows down the training). We fix this so that in the self-attention (our case), only one linear module is needed and thus speed up out training, which has been proved in out testing (posted in the internal diff)

Reply:
BTW, I am a little confused about the test failure. Seems like it fails in test_adam and some error msg like:
test_fixed_cuda_assert_async (__main__.TestCuda) ... C:\actions-runner\_work\pytorch\pytorch\aten\src\ATen\native\cuda\TensorCompare.cu:144: block: [0,0,0], thread: [0,0,0] Assertion `input[0] != 0` failed.
2022-02-18T02:48:23.8845865Z Traceback (most recent call last):
2022-02-18T02:48:23.8847495Z   File "<string>", line 4, in <module>
2022-02-18T02:48:23.8848545Z   File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\cuda\__init__.py", line 495, in synchronize
2022-02-18T02:48:23.8849408Z     return torch._C._cuda_synchronize()
2022-02-18T02:48:23.8850077Z RuntimeError: CUDA error: device-side assert triggered

Full logs are seen in https://pipelines.actions.githubusercontent.com/mBh68xKhi8LyM7tp3vECvYXNFvuV4gyVGgmYCteuEZP9JH92QN/_apis/pipelines/1/runs/1769331/signedlogcontent/27?urlExpires=2022-02-18T06%3A04%3A12.7207992Z&urlSigningMethod=HMACV1&urlSignature=gev8b7dbjy9NngHpNRP%2BqHFdUO0FYESAhrkpFWxTY2s%3D. I don't see the relationship between my change and these errors. Could you give any hints about this? Thanks.

Reply:
Thanks for the details.
I think that you want to update the comments in the code to reflect the fact that this whole logic aims at preserving the "is" property between the three values. And it is not really important what each case corresponds to,

Reply:
Yeah thanks for the great suggestion! Let me update the comments to avoid confusion.